src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -0,0 +1,201 @@
+
+        /// <summary>
+        /// Constructs the <see cref="ReversedDiagnosticsServer"/> instance with an endpoint bound
+        /// to the location specified by <paramref name="transportPath"/>.
+        /// </summary>
+        /// <param name="transportPath">
+        /// The path of the server endpoint.
+        /// On Windows, this can be a full pipe path or the name without the "\\.\pipe\" prefix.
+        /// On all other systems, this must be the full file path of the socket.
+        /// </param>
+        public ReversedDiagnosticsServer(string transportPath)
+            : this(transportPath, MaxAllowedConnections)
+        {
+        }
+
+        /// <summary>
+        /// Constructs the <see cref="ReversedDiagnosticsServer"/> instance with an endpoint bound
+        /// to the location specified by <paramref name="transportPath"/>.
+        /// </summary>
+        /// <param name="transportPath">
+        /// The path of the server endpoint.
+        /// On Windows, this can be a full pipe path or the name without the "\\.\pipe\" prefix.
+        /// On all other systems, this must be the full file path of the socket.
+        /// </param>
+        /// <param name="maxConnections">The maximum number of connections the server will support.</param>
+        public ReversedDiagnosticsServer(string transportPath, int maxConnections)
+        {
+            _transport = IpcServerTransport.Create(transportPath, maxConnections);
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _cancellation.Cancel();
+
+                IEnumerable<ServerIpcEndpoint> endpoints = _endpoints.Values;
+                _endpoints.Clear();
+
+                foreach (var endpoint in endpoints)
+                {
+                    endpoint.Dispose();
+                }
+
+                _transport.Dispose();
+
+                _cancellation.Dispose();
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        /// Provides connection information when a new runtime instance connects to the server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>A <see cref="ReversedDiagnosticsConnection"/> that contains information about the new runtime instance connection.</returns>
+        /// <remarks>
+        /// This will only provide connection information on the first time a runtime connects to the server. Subsequent
+        /// reconects will update the existing <see cref="ReversedDiagnosticsConnection"/> instance. If a connection is removed
+        /// using <see cref="RemoveConnection(Guid)"/> and the same runtime instance reconnects afte this call, then a
+        /// new <see cref="ReversedDiagnosticsConnection"/> will be produced.
+        /// </remarks>
+        public async Task<ReversedDiagnosticsConnection> AcceptAsync(CancellationToken token)
```

Comment:
This API feels misleading. The signature and comments suggest that if I didn't care about getting any more clients then I wouldn't need to call this any longer, similar to socket.accept(). Of course if I tried to implement that all my existing clients would stop responding to new commands too.  I am hoping we can rename and potentially refactor this so people understand they have to commit to keep this loop pumping. For example Task ProcessConnectionsAsync(token) and the new connections get reported via some callback or event?
I also think filtering out clients we've seen before is unnecessarily dragging this type into client management and I'm not sure it provides much value for the consumers who already have to do client management anyways.

Reply:
If the server class doesn't filter existing connections, that pushes the filtering onto the caller. I was trying to enable the what was briefly discussed at https://github.com/dotnet/diagnostics/pull/931#discussion_r423459697.

Otherwise, the caller would have to implement something like:
```
var connections = new Dictionary<Guid, Connection>();
var server = new ReversedDiagnosticsServer(address);
while (true)
{ 
    var connection = await server.ProcessConnectionsAsync(token);
    if (!connections.TryGetValue(connection.RuntineInstanceIdentifier, out _))
    {
       _ = HandleConnectionAsync(connection);
    }
}
```
My understanding is that the caller is unlikely going to want to handle a new connection from an existing runtime instance the same way that it handles a new connection from a new runtime instance.

There would be additional locking if the caller wants to be multithreaded in order to dole out connections simultaneously.

From the same comment, it seemed like there was a desire to avoid using callbacks or events for providing the connections. If we were to go the route of using those, I would think that we'd use something like:

```
var connections = new Dictionary<Guid, Connection>();
var server = new ReversedDiagnosticsServer(address);
server.Listen(connection =>
{
    if (!connections.TryGetValue(connection.RuntineInstanceIdentifier, out _))
    {
       _ = HandleConnectionAsync(connection);
    }
});
```

Reply:
What if I split the current implementation of AcceptAsync into two methods:

1) `void Listen(int maxConnections)`

which creates the server endpoint, loops to accept connections, and puts new connections into a queue. The method is non-blocking; internally it would create a task that would asynchronously accept connections.

The new connections are then consumed by

2) `Task<IpcEndpointInfo> AcceptAsync(CancellationToken token)`

which pulls new connections off of the queue, tests that the endpoint is valid (because it could be sitting in the queue for a while depending on how often the consumer calls AcceptAsync, especially if a runtime instance connects and then shutdowns down before AcceptAsync pulls it off the queue), and returns the endpoint.

This separates the need to 'pump' the server loop from whatever behavior the caller of the AcceptAsync wants to do (accept only one connection and be done, loop through connections as they become available, etc).

This can also allow us to effectively shutdown the server endpoint and restart it at later time without requiring the caller to cancel AcceptAsync, dispose the server, etc, if we want to enable that behavior at some point.